src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,140 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail == null) {
+            head = newNode;
+        } else {
+            tail.setNext(newNode);
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> newNode = new Node<>(current.getPrev(), value, current);
+            if (current.getPrev() == null) {
+                head = newNode;
+            } else {
+                current.getPrev().setNext(newNode);
+            }
+            current.setPrev(newNode);
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).getValue();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNode(index);
+        T oldValue = current.getValue();
+        current.setValue(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNode(index);
+        T value = current.getValue();
+        unlink(current);
```

Comment:
Consider creating a separate method `unlink` that handles the logic of unlinking nodes. This method should be used here and in the `remove(T object)` method to avoid code duplication.